### PR TITLE
feat: add InputLengthGuard and OutputLengthGuard (#1053)

### PR DESCRIPTION
**Pull request by vibeyclaw**
Closes #1053

## What this PR does

Implements character-based length guardrails for both LLM input and output, as described in the issue.

### New classes

**`InputLengthGuard`** — blocks requests where the total character count across all messages in the history exceeds `max_chars`:

```python
import railtracks as rt

input_guard = rt.guardrails.llm.InputLengthGuard(max_chars=4000)
```

**`OutputLengthGuard`** — blocks responses where the assistant reply exceeds `max_chars`:

```python
output_guard = rt.guardrails.llm.OutputLengthGuard(max_chars=2000)
```

### Design decisions

- **Characters as the unit** (as recommended in the issue) — no external tokenizer dependency, easy to reason about, fast
- Both guards raise `ValueError` on non-positive `max_chars` to catch misconfiguration early
- Decisions include a `meta` dict with `total_chars` and `max_chars` for observability/logging
- Both guards are exposed at `railtracks.guardrails.llm.*` for consistent import patterns

### Files changed

| File | Change |
|------|--------|
| `guardrails/llm/input/length_guard.py` | New — InputLengthGuard implementation |
| `guardrails/llm/output/length_guard.py` | New — OutputLengthGuard implementation |
| `guardrails/llm/input/__init__.py` | Export InputLengthGuard |
| `guardrails/llm/output/__init__.py` | Export OutputLengthGuard |
| `guardrails/llm/__init__.py` | Re-export both guards at package level |
| `tests/unit_tests/guardrails/test_length_guards.py` | 20 unit tests covering allow/block boundary, edge cases, error handling |

### Future work (out of scope for this PR)
- Word-based counting (`max_words`)
- Token-based counting via optional tokenizer (tiktoken / transformers)

### CODE REVIEW

File: uv.lock [hunk 1]
```diff
@@ -158,18 +158,6 @@ wheels = [
     { url = "https://files.pythonhosted.org/packages/6f/12/e5e0282d673bb9746bacfb6e2dba8719989d3660cdb2ea79aee9a9651afb/anyio-4.10.0-py3-none-any.whl", hash = "sha256:60e474ac86736bbfd6f210f7a61218939c318f43f9972497381f1c5e930ed3d1", size = 107213, upload-time = "2025-08-04T08:54:24.882Z" },
```
Comment (Amir-R25) [inline uv.lock:158]: Please restore the `uv.lock` file, we'll handle the required updates on our side

**Review by Amir-R25 — CHANGES_REQUESTED** [review]
Hi,

Thanks for the PR! This will be a useful feature to limit costs for agents!

I added my comment regarding the uv.lock before I can approve running of the workflow tests.

Could you please add relevant documentation as well alongside these two guardrails? in `docs/guardrails/`, let's call it `character_guards.md`

**Review by Pooria90 — CHANGES_REQUESTED** [review]
Looks good to me!

Only the `uv.lock` needs to be reverted.

**Comment by Amir-R25** [comment]
@vibeyclaw checking in here again on this PR, thanks

**Comment by Amir-R25** [comment]
@vibeyclaw checking in again, thanks!

**Comment by Amir-R25** [comment]
Hey @vibeyclaw, love the work here.

Just need a small change before we can merge. If we don't hear back in a few days we'll apply the fix ourselves and credit you. 

Thanks!

**Comment by vibeyclaw** [comment]
Done! I've reverted `uv.lock` back to the main branch state. Thanks for the patience and the follow-ups, @Amir-R25 — sorry for the delay!

**Comment by Amir-R25** [comment]
@vibeyclaw Thanks, would you be able to update your branch with main as well, there are some conflicts. Alternatively I can do it if not possible

image

**Comment by Amir-R25** [comment]
Hey @vibeyclaw, 

this PR has been open for over 30 days and we'd love to get it merged. The conflicts are only in 3 files, all straightforward. Just replace their contents with the following and push:

**`packages/railtracks/src/railtracks/guardrails/llm/__init__.py`**
```python
from . import input, output
from ._pii.config import PIICustomPattern, PIIEntity, PIIRedactConfig
from .input.block_text import BlockTextInputGuard
from .input.length_guard import InputLengthGuard
from .input.pii_redact import PIIRedactInputGuard
from .mixin import LLMGuardrailsMixin
from .output.block_text import BlockTextOutputGuard
from .output.length_guard import OutputLengthGuard
from .output.pii_redact import PIIRedactOutputGuard

__all__ = [
    "input",
    "output",
    "BlockTextInputGuard",
    "BlockTextOutputGuard",
    "InputLengthGuard",
    "LLMGuardrailsMixin",
    "OutputLengthGuard",
    "PIICustomPattern",
    "PIIEntity",
    "PIIRedactConfig",
    "PIIRedactInputGuard",
    "PIIRedactOutputGuard",
]
```

**`packages/railtracks/src/railtracks/guardrails/llm/input/__init__.py`**
```python
from .block_text import BlockTextInputGuard
from .length_guard import InputLengthGuard
from .pii_redact import PIIRedactInputGuard

__all__ = ["BlockTextInputGuard", "InputLengthGuard", "PIIRedactInputGuard"]
```

**`packages/railtracks/src/railtracks/guardrails/llm/output/__init__.py`**
```python
from .block_text import BlockTextOutputGuard
from .length_guard import OutputLengthGuard
from .pii_redact import PIIRedactOutputGuard

__all__ = ["BlockTextOutputGuard", "OutputLengthGuard", "PIIRedactOutputGuard"]
```

Then run `git add` on those 3 files, commit, and push, that should clear the conflicts and we can merge. Thanks!

**Comment by Amir-R25** [comment]
@vibeyclaw Unfortunately we will have to close this PR as it has been open for more than 30 days. We appreciate your contribution and have launched a new PR using your commits to credit your work: #1091 so we can handle the merge conflicts
